### DISCUSSION TITLE
Add support for Enum fields

### DISCUSSION THREAD

**Discussion by yellowBanano** [directus/directus#4887] [Feature Requests]
Not allow to change or work with enum type from MySQL 8. In order to change smth for field, I need to modify settings directly in database in `directus_fields` table

My version of directus --  Directus 9.0.0-rc.51 
DB -- MySQL 8
deployment -- npx

![изображение](https://user-images.githubusercontent.com/32203047/112490584-8c162280-8d90-11eb-99a3-10eae887cd2d.png)
![изображение](https://user-images.githubusercontent.com/32203047/112490661-a05a1f80-8d90-11eb-80ea-aa283732a8e5.png)

**Comment by benhaynes** [comment]
We'll need to look into SQL vendor support for this, but I imagine the process of adding this to be:

* Add support for the **type** in our Directus super-set data type list
* Add a conditional field to the Field's `Schema` pane for adding/managing the list of permitted values
* Update the applicable interfaces (dropdown, radios, etc) to work with this type, pulling options from the schema instead of the `choices` interface options

**Reply by delucca** [reply]
@benhaynes first and foremost, thanks for this impressive tool :) I've been searching for something similar to this for ages!

Regarding this discussion, is ENUM types supported already? I have a few ENUMs in my Postgres database and I would love them to be simple dropdown lists.

**Reply by benhaynes** [reply]
Thank you so much... that means a lot. ❤️ 

I don't think we have support for enum as of now... we're trying to figure out "choices" within other datatypes. So that's the current thinking/delay.

**Reply by delucca** [reply]
Oh, I see! Thanks for your feedback.

Just a quick question, right now when I have an ENUM type I can't even set the interface. Is it possible to allow setting the interface manually? For example, I could use the "Dropdown" input type (that is available when the field is a string) to manually define the enum values.

This is possible? Or, there are any workaround for this?

**Reply by benhaynes** [reply]
That I'm unsure of — @rijkvanzanten, are you aware of how we handle enums as of now?

**Reply by rijkvanzanten** [reply]
We don't have any special handling for enum fields. I believe some database vendors might return those choices as part of the default value definition, others as part of the type. As of right now, Directus' dropdown interface doesn't work with ENUM types in the database 👍🏻

**Reply by delucca** [reply]
Thanks for the info.

I'll eagerly wait for this support :)

**Reply by minyan-gu** [reply]
hey guys - just check if there is any update on this. 

For the field type detected as `unknown`, please can we at least set up our own dropdown list values through the interface, the values doesn't have to get sync'd with the ENUM type from DB to start with (ofc ideally should be). We can start off by setting those dropdown values manually, which is a great improvement already.

On latest version of Directus, the interface is not available still for unknown type, so the user has to enter those database enum values manually, which is quite a pain.

**Reply by BenStirrup** [reply]
The suggestion of @minyan-gu seems sensible. Would it be possible to do it any way or with a workaround ?

**Comment by TyrionFront** [comment]
HI team !)
Just another one "thumbs up" to make this feature alive.

**Reply by SharadKumar** [reply]
Prisma supports enums. I am using it on a project with Directus, but not able to utilise enum for things like `status` fields.

I would be quite useful if Directus had a way to support this (or even a workaround)! Not always can we use Directus with a new database.

**Reply by TyrionFront** [reply]
I'm currently using Drizzle-ORM which supports it as well and in fact in most cases is a way more performant than Prisma. But the point is not about using enums within a codebase - on that lvl you may have all required flexibility you need. it's about the ability to work with enums within DIrectus UI and that is not an option yet, unfortunately.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by BenStirrup** [comment]
Any update ? 🙏 The absence of such a feature is a deal breaker for us 😢.

**Comment by ihadabs** [comment]
@rijkvanzanten any update on this?
At least allowing the choosing of any interface on Directus if the `type=unknown`
This might also work help for all unknown types.

Currently, it shows no options for interface or display:

<img width="829" alt="CleanShot 2023-10-01 at 19 54 04@2x" src="https://github.com/directus/directus/assets/34512743/0a0b10bb-0e2d-485e-9149-385f2703fb35">

<img width="845" alt="CleanShot 2023-10-01 at 19 54 14@2x" src="https://github.com/directus/directus/assets/34512743/9bba27a9-df34-4b69-b6a3-aaab406d7c64">

**Reply by rijkvanzanten** [reply]
I'm down to discuss adding a new "raw" interface that can be used with fields of unknown types, but I don't think that blanketly allowing just the select-dropdown for any unknown field is the right solution for supporting Enums. For enums specifically, I'm also not sure if the existing dropdown interface is the right interface for the job. I'd expect the interface to pull the available options from the database itself rather than a duplicate set of options on the interface level 🤔

**Reply by ihadabs** [reply]
@rijkvanzanten Pulling available enum options would be great, but still, for us and most teams it is a one-time setup so we would not mind if there was a drop-drown  interface and we could add the options ourselves ^^

If we'll build something just for enums, then it's basically a dropdown with all of its configuration BUT the options values have to match one of the enum values.  Also, options have to be reevaluated whenever Directus fetches the schema and discovers that some enum values have been changed. (Directus can still show the dropdown options that have a matching enum value).

**Example:**
```
enum = "green" | "red"
```
```
directus options:
title: "Some Green", value: "green"
title: "The red", value: "red"
```

if enum changed to = `"green" | "blue"`
```
directus options:
title: "Some Green", value: "green"
title: "The red", value: "red" /// This should not be displayed for editors, and maybe show an error in the field configuration settings

/// Directus can maybe also add the default shape for an enum value
/// Where title is same as the value
title: "blue", value: "blue"

/// So basically Directus will fetch enum values and while setting up the field
/// + it will add the options which are missing (by checking current options values if there were any)
/// + it may delete any options which don't have a matching enum value
```

**Comment by SharadKumar** [comment]
Just having dropdown Interface to handle enums, that Admin can once setup, will be quite beneficial than not having any support at all.

**Reply by SharadKumar** [reply]
This issue was raised for v9. I can confirm it is also applicable to the latest v10.9.x

**Comment by EdouardDem** [comment]
@SharadKumar 
Hi everyone. I've found a workaround for a specific use case using the `directus-sync` CLI. Here's how you can do it:

1. First, set up the Directus-Sync CLI as outlined here: [directus-sync GitHub](https://github.com/tractr/directus-sync).
2. Dump the structure of your Directus instance by running `npx directus-sync pull`. This command leverages [Directus's snapshot endpoint](https://docs.directus.io/reference/system/schema.html#retrieve-schema-snapshot) to save your Directus configuration into separate JSON files.
3. Locate the file representing the "unknown" field, such as a Postgres Enum field, e.g., `directus-config/snapshot/fields/organization/type.json`. Edit the JSON to "force" the dropdown interface by modifying the `meta.interface` and `meta.options` properties. For example:

```json
{
  "collection": "organization",
  "field": "type",
  "type": "unknown",
  "meta": {
    "collection": "organization",
    "conditions": null,
    "display": "raw",
    "display_options": null,
    "field": "type",
    "group": null,
    "hidden": false,
    "interface": "select-dropdown",
    "note": null,
    "options": {
      "choices": [
        {
          "value": "startup",
          "text": "Startup"
        },
        {
          "value": "incubator",
          "text": "Incubator"
        },
        {
          "value": "enterprise",
          "text": "Enterprise"
        }
      ]
    },
    "readonly": false,
    "required": true,
    "sort": 4,
    "special": null,
    "translations": null,
    "validation": null,
    "validation_message": null,
    "width": "half"
  },
  "schema": {
    "name": "type",
    "table": "organization",
    "data_type": "\"OrganizationType\"",
    "default_value": null,
    "max_length": null,
    "numeric_precision": null,
    "numeric_scale": null,
    "is_nullable": false,
    "is_unique": false,
    "is_primary_key": false,
    "is_generated": false,
    "generation_expression": null,
    "has_auto_increment": false,
    "foreign_key_table": null,
    "foreign_key_column": null
  }
}
```

4. Push the modified schema back to Directus with `npx directus-sync push`.
5. You should now see the dropdown in the form and also in the search bars.

**Notes:**

Currently, I haven't been able to implement this with a multi-select dropdown. In Postgres, the format for a list of enums is `{value1,value2,value3}`. Directus doesn't natively parse and stringify this format. A custom transformer, like `cast-csv`, could potentially be incorporated here: [Directus Payload Service](https://github.com/directus/directus/blob/main/api/src/services/payload.ts#L130-L145).

I hope this solution is helpful.

**Comment by mpcref** [comment]
I would like to be able to group by enum fields on the Kanban Layout.
Is there any way I can work around the current limitation in order to achieve this?

**Comment by Davut7** [comment]
Any news? Is this feature added?

**Reply by rijkvanzanten** [reply]
No, it's an open feature request. Enums as a concept aren't supported as a type in the SQL standard, which means adding support for it in Directus gets a bit trickier. We currently don't have a concept of 'supports' flags for each database driver, so there's no opt-in for things like enums for case-by-case database drivers

**Comment by nikelborm** [comment]
Not sure if i can call it workaround, given the amount of code and time it took, but here's mine. You just connect to the same postgres instance, where Directus lives, and run this command. It will discover all tables in the entire database with any enum columns and create/update according records in `directus_fields`. Feel free to use it, but remember that if you didn't make preliminary backup, I'm not responsible for that 😜 

```sql
with
  enums as (
    select
      pg_type.oid as enum_id,
      any_value(pg_type.typname) as enum_name,
      array_agg(pg_enum.oid) as enum_value_id,
      array_agg(pg_enum.enumlabel) as enum_value,
      any_value(pg_type.typnamespace) as schema_id,
      any_value(pg_namespace.nspname) as schema_name,
      json_build_object(
        'choices',
        json_agg(
          json_build_object(
            'text', replace(pg_enum.enumlabel::text, '_', ' '),
            'value', pg_enum.enumlabel
          )
        )
      ) as options
    from pg_type
      join pg_enum on pg_type.oid = pg_enum.enumtypid
      join pg_catalog.pg_namespace ON pg_namespace.oid = pg_type.typnamespace
    GROUP BY enum_id
  ),
  table_columns as (
    SELECT
      table_schema as table_schema_name,
      table_name,
      is_nullable,
      column_name,
      schema_name as enum_schema_name,
      enum_name,
      enums.options as enum_options
    FROM information_schema.columns
    inner join enums on (
      columns.udt_name = enums.enum_name
    ) AND (
      columns.udt_schema = enums.schema_name
    )
  )
MERGE INTO directus_fields
USING table_columns
ON collection = table_name AND field = column_name
WHEN MATCHED THEN
  UPDATE SET
    required = (is_nullable = 'NO'),
    interface = 'select-dropdown',
    display = 'labels',
    options = enum_options,
    display_options = enum_options
WHEN NOT MATCHED THEN
  INSERT (
    collection,
    field,
    required,
    interface,
    display,
    options,
    display_options
  )
  VALUES (
    table_name,
    column_name,
    is_nullable = 'NO',
    'select-dropdown',
    'labels',
    enum_options,
    enum_options
  );
```

I didn't bother to write deep merge of old `display_options` and `options` with their new versions, which means that rerunning this script after you added some styles to your enum values in Directus UI such as colors or icons will reset those styles to default values. But if anybody cares about this enough, I'm happy to help

**Reply by nikelborm** [reply]
Turns out that sql I provided here is quite not enough to call it done. Since first showing it here, i was able to gain much more improvement in how enums are handled in directus, I'll send it later here

**Comment by cdwmhcc** [comment]
mv/view/enum are all frequently used, but unfortunately Directus does not support them at present. It is indeed a big technical challenge.

**Comment by souserge** [comment]
I see the difficulty with maintaining different SQL drivers, but would it be at least possible to "hack" the enum in the GraphQL API somehow? This is a significant constraint trying to write a type-safe frontend consuming Directus API.

**Comment by danielbuechele** [comment]
Suggesting a minimal solution to the problem in https://github.com/directus/directus/pull/26811
